### DISCUSSION TITLE
Monogame integration?

### DISCUSSION THREAD

**Discussion by vrgamestudio** [StereoKit/StereoKit#131] [General]
I've just found this very interesting project you have here and am wondering if you have any information regards how this could be integrated with Monogame projects, even if it's just on desktop for now (+android later)?

I believe the latest version of Monogame (3.8) does use .NET Core now so should be compatible with this but would that be more hassle than it's worth given it must have already crossed your mind (this is also a C# graphics/game framework) yet there is no mention of it so I'm thinking maybe there are fundamental reasons for it not being a good fit, so was better to start over again? 

It would be lovely if they could play nice together though! Could it be made to work and if so, maybe a little tutorial on how?

**Comment by maluoi** [comment]
I used to work with XNA, way back before Unity was a thing! It was absolutely a big inspiration to me. While I haven't used MonoGame or tracked its development, I do love that MonoGame has still kept the XNA legacy alive :)

Unfortunately, I would describe MonoGame as an alternative to StereoKit rather than something that complements it. As you mentioned, it is indeed also a C# graphics framework, and the core responsibilities have significant overlap!

If you have feature requests for StereoKit to bring in functionality that MonoGame already has, I'd be happy to take a look and consider it!

**Reply by vrgamestudio** [reply]
Having previously been interested in Managed DirectX, I too followed and used XNA and remained a fan up until Microsoft pulled the plug on it. 

Until last year I'd considered XNA dead and my XNA projects dead also, they existed purely in backups and old videos on youtube. 

Then Monogame once again came to my attention given they now had the content pipeline tool up and running and suddenly I found that against all odds XNA really did live once again, was mature and now was multiplatform too. I take my hat off to the dedicated people who made that happen.

One thing Monogame lacks however is anything at all in the direction of Virtual Reality, so what I was looking for was a library (C# wrapper) around the recent OpenXR libraries that vendors have agreed to make for their systems. I've been looking around and unfortunately found nothing. I don't feel I have the time or the talent to write my own and it would be so very speculative a move to make at this time. Then I saw StereoKit.

I haven't looked very much at StereoKit, I only learned of it's existence yesterday but it was already clear there was alot of overlap with Monogame, which is great but all I wanted was the OpenXR part, to be able to integrate VR into existing Monogame applications via OpenXR. 

My argument to myself would be that popularity and stability count for alot as does already invested time and learning. Xna, now Monogame is one of the rare birds that actually got alot of traction, has acclaimed shipped games and is the real deal so I would prefer to at least have the option to stay with it or slowly migrate parts of the code over to StereoKit rather than convert 100% to StereoKit simply for the VR functionality which, in the scheme of market share etc is still tiny right now. 

So while I would not like to discourage you in any way to bring in Xna like functionality to StereoKit, to me it's like reinventing the wheel. Maybe the wheel needs reinventing, maybe you can do a better job at implementing a game framework; after all Xna was released 15 years ago and times change and Monogame is by no means perfect or without problems and issues but game dev is hard and fragmentation makes things harder still, hence the need for OpenXR in the first place.

What would delight me at this point would be if your OpenXR wrappers could be separated from your game framework and used standalone, or with Monogame Matrix structs etc, with your (what looks very nifty so far but I have no clue about it) game framework there as a great option for non Xna people coming fresh just wanting to dive in.

Having said all this I am extremely interested in StereoKit, it looks like a very impressive piece of work and I want to see more of it. Since you've made the source code available I guess it wouldn't be impossible for someone like myself to come in and split the source code to make the thing I've described based on your hard work but then I've just dropped in here from10,000 feet and know nothing about it so guidance and help in doing that would be preferable.

I suppose in a nutshell I'm saying that I think StereoKit and Monogame should Team up and make something truly great for the future :)

**Reply by maluoi** [reply]
MonoGame was actually something I considered using when I first started building StereoKit, and ultimately decided not to. It's possible that I could have designed StereoKit to be a layer over MonoGame, but that didn't jive too well with my goals either.

Here's a a few of the major reasons why I chose not to do anything with MonoGame:
1. StereoKit is a higher level API than XNA. I try my best to cater to people that are not good at, or not interested in math, graphics, or 3D content in general, and so I provide many features that simplify those tasks even further than XNA did.
2. StereoKit is not specifically for games. XR is important for regular apps and tools, and if games are no longer client no. 1 (like they are in game engines), then there are a number of things that can and should be done differently. In particular, my lighting and rendering are aggressively tuned for simplicity and mobile/VR performance, and the asset pipeline is designed for "bake free" usage, where StereoKit will optimize user friendly asset format at runtime rather than at compile time.
3. StereoKit prioritizes XR above all else! This means that flat-screen is treated as a development environment only, non-vr hardware is excluded from compatibility testing, certain newer hardware features are assumed, and systems like input are entirely designed around XR input scenarios. This is much better for the XR developer since they can assume the defaults are correct for XR, and will struggle less to do XR tasks. This is possibly a rather forward looking decision, but I'd rather be early instead of late.
4. StereoKit is at its core a C API, which allows it to be exposed to other languages like Rust, or compiled to WebAssembly. I love C#, and consider it to be the primary distribution vector for StereoKit, but I'm a big supporter of choosing your own language for development! There are in fact a number of StereoKit developers that are working from the C++ side already :)
5. MonoGame had (has?) a dependency on SharpDX, which is no longer maintained. XR is on the bleeding edge, and I wanted to make sure I'd have easy access to modern XR oriented graphics extensions if I needed them.

I still think it'd be great for MonoGame to have some tie-ins for OpenXR, and I'd be happy to offer advice and reference code to anyone that tries to do it, but I'm afraid it's outside of my own current priorities :)

**Reply by TopperDEL** [reply]
Interesting discussion! I just want to hint @vrgamestudio to another game engine: WaveEngine. I don't want to push you away from StereoKit, but maybe that one solves your current needs, if you don't know it yet.

I'm using StereoKit for a productive app and not for a game and I heavily appreciate those little things that make my live as a "top-level-programmer" easier by using SK.
